provider/apple.go

```diff
@@ -0,0 +1,534 @@
+	KeyID              string   // private key ID  assigned to private key obtain in Apple developer account
+	UserAgent          string   // UserAgent value for Apple API request. Default: "github.com/go-pkgz/auth"
+	ClientSecretExpire int64    // time in seconds when client secret will expired (default: 24h)
+	Scopes             []string // for apple provider allow only "email" and "name" scope values
+
+	privateKey   interface{} // private key from Apple obtained in developer account (the keys section). Required for create the Client Secret (https://developer.apple.com/documentation/sign_in_with_apple/generate_and_validate_tokens#3262048)
+	clientSecret string      // is the JWT client secret will create after first call and then used until expired
+}
+
+// AppleHandler implements login via Apple ID
+type AppleHandler struct {
+	Params
+
+	// all of these fields specific to particular oauth2 provider
+	name string
+	// infoURL  string not implemented at Apple side
+	endpoint oauth2.Endpoint
+
+	mapUser func(jwt.MapClaims) token.User // map info from InfoURL to User
+	conf    AppleConfig                    // main config for Apple auth provider
+
+	PrivateKeyLoader PrivateKeyLoaderInterface // custom function interface for load private key
+
+}
+
+// PrivateKeyLoaderInterface interface for implement custom loader for Apple private key from user source
+type PrivateKeyLoaderInterface interface {
+	LoadPrivateKey() ([]byte, error)
+}
+
+// LoadFromFileFunc is the type for use pre-defined private key loader function
+// Path field must be set with actual path to private key file
+type LoadFromFileFunc struct {
+	Path string
+}
+
+// LoadApplePrivateKeyFromFile return instance for pre-defined loader function from local file
+func LoadApplePrivateKeyFromFile(path string) LoadFromFileFunc {
+	return LoadFromFileFunc{
+		Path: path,
+	}
+}
+
+// LoadPrivateKey implement pre-defined (built-in) PrivateKeyLoaderInterface interface method for load private key from local file
+func (lf LoadFromFileFunc) LoadPrivateKey() ([]byte, error) {
+	if lf.Path == "" {
+		return nil, errors.New("empty private key path not allowed")
+	}
+
+	kFile, err := os.Open(lf.Path)
+	if err != nil {
+		return nil, err
+	}
+	keyValue, err := ioutil.ReadAll(kFile)
+	if err != nil {
+		return nil, err
+	}
+	err = kFile.Close()
+	return keyValue, err
+}
+
+// NewApple create new AppleProvider instance with custom name and parameters
+// endpoints used in constructor  required only for testing and can't be define one outside the package.
+// Private key must be set, when instance create call, for create `client_secret`
+func NewApple(name string, p Params, appleCfg AppleConfig, endpoints oauth2.Endpoint, privateKeyLoader PrivateKeyLoaderInterface) (*AppleHandler, error) {
+
+	if name == "" {
+		return nil, errors.New("empty name for apple provider not allowed")
+	}
+
+	if p.L == nil {
+		p.L = logger.NoOp
+	}
+
+	// configuring default values
```

Comment:
are those params should be even configurable? Maybe we just use const here as I don't really see the reason to redefine user agent, auth url and token url

Reply:
I'm agree with you about exclude user-agent from config params, but auth and token urls not include to `AppleConfig` and exist only in constructor `NewApple`. It's need for change they for tests only.

Reply:
I would rather change it directly in tests if this is the only reason for the exposure if we are talking about same-package unit tests only

Reply:
Ok, will change it